hswidget/fonttablewidget/widget.go

```diff
@@ -0,0 +1,310 @@
+	upW, upH                 = 30, 30
+	downW, downH             = 40, 30
+	addW, addH               = 400, 30
+	editRuneW, editRuneH     = 50, 30
+	saveCancelW, saveCancelH = 80, 30
+)
+
+type textures struct {
+	up, down, del *giu.Texture
+}
+
+// widget represents a font table widget
+type widget struct {
+	fontTable *d2font.Font
+	id        string
+	textures
+}
+
+// Create creates a new FontTable widget
+func Create(
+	up, down, del *giu.Texture,
+	id string, fontTable *d2font.Font,
+) giu.Widget {
+	result := &widget{
+		fontTable: fontTable,
+		id:        id,
+		textures: textures{
+			up, down, del,
+		},
+	}
+
+	return result
+}
+
+// Build builds a widget
+func (p *widget) Build() {
+	state := p.getState()
+
+	switch state.mode {
+	case fontTableWidgetViewer:
+		p.makeTableLayout().Build()
+	case fontTableWidgetEditRune:
+		p.makeEditRuneLayout().Build()
+	case fontTableWidgetAddItem:
+		p.makeAddItemLayout().Build()
+	}
+}
+
+func (p *widget) makeTableLayout() giu.Layout {
+	state := p.getState()
+
+	rows := make(giu.Rows, 0)
+
+	rows = append(rows, giu.Row(
+		giu.Label("Delete"),
+		giu.Label("Index"),
+		giu.Label("Character"),
+		giu.Label("Width (px)"),
+	))
+
+	// we need to get keys from map[rune]*d2font.fontGlyph
+	// and then sort them
+	chars := make([]rune, len(p.fontTable.Glyphs))
+
+	// reading runes from map
+	idx := 0
+
+	for r := range p.fontTable.Glyphs {
+		chars[idx] = r
+		idx++
+	}
+
+	// sorting runes
+	sort.Slice(chars, func(i, j int) bool {
+		return p.fontTable.Glyphs[chars[i]].FrameIndex() < p.fontTable.Glyphs[chars[j]].FrameIndex()
+	})
+
+	for _, idx := range chars {
+		rows = append(rows, p.makeGlyphLayout(idx))
+	}
+
+	return giu.Layout{
+		giu.Child("##" + p.id + "tableArea").Border(false).Layout(giu.Layout{
+			giu.FastTable("##" + p.id + "table").Border(true).Rows(rows),
+			giu.Separator(),
+			giu.Button("add##"+p.id+"addItem").Size(addW, addH).OnClick(func() {
+				state.mode = fontTableWidgetAddItem
+			}),
+		}),
+	}
+}
+
+func (p *widget) makeGlyphLayout(r rune) *giu.RowWidget {
+	state := p.getState()
+
+	if p.fontTable.Glyphs[r] == nil {
+		return &giu.RowWidget{}
+	}
+
+	w := p.fontTable.Glyphs[r].Width()
+	width32 := int32(w)
+
+	row := giu.Row(
+		giu.Line(
+			// in fact, it should be ImageButton, but this shit doesn't work
+			// (imgui.PushID returns panic) :-/
+			giu.Button("del##"+p.id+"deleteFrame"+string(r)).Size(delW, delH).OnClick(func() {
+				p.deleteRow(r)
+			}),
+		),
+		giu.Line(
+			giu.Label(fmt.Sprintf("%d", p.fontTable.Glyphs[r].FrameIndex())),
+			giu.Button("up##"+p.id+"upItem"+string(r)).Size(upW, upH).OnClick(func() {
+				p.itemUp(r)
+			}),
+			giu.Button("down##"+p.id+"upItem"+string(r)).Size(downW, downH).OnClick(func() {
+				p.itemDown(r)
+			}),
+		),
+		giu.Line(
+			giu.Button("edit##"+p.id+"editRune"+string(r)).Size(editRuneW, editRuneH).OnClick(func() {
+				state.editRune.startRune = r
+				state.mode = fontTableWidgetEditRune
+			}),
```

Comment:
:confused: there should be `ImageButtons` but hsutil.MakeImageButton causes panic:
```golang
panic: Assertion failed!
File: /home/user/go/pkg/mod/github.com/ianling/imgui-go@v1.12.1-0.20210115050955-434395a2da82/imgui/imgui.h, Line 1421

Expression: Size > 0
```
exactly it is caused by imgui.PushID()

Reply:
have you tried running the debugger to see what is causing the panic?

Reply:
I can't see absolutly nothing, what could cause this crash. in cof editor works fine, but here...

Reply:
but I'm working on workaround

Reply:
a95c37f